### PR TITLE
Docs - CDI guide - introduce the "Non-standard Features" section

### PR DESCRIPTION
**Pull request by mkouba**
- also try to clarify "Native Executables and Private Members" section

### CODE REVIEW

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 1]
```diff
@@ -168,6 +170,155 @@ public class CounterBean {
...
+public class CustomTracerConfiguration {
+
+    @Produces
+    public Reporter reporter(){
+        // create a custom Reporter
+    }
+}
+----
+
+`@DefaultBean` allows extensions (or any other code for that matter) to provide defaults while backing off if beans of that type are supplied in any
```
Comment (manovotn) [inline docs/src/main/asciidoc/cdi-reference.adoc]: This is a repetition of the above I think it could removed.

**Review by manovotn — COMMENTED** [review]
Added few remarks, some of which belong to the docs that were already there. But you know, since you starting poking into it... ;-)

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 2]
```diff
@@ -168,6 +170,155 @@ public class CounterBean {
 * Interceptor methods on superclasses are not implemented yet
 * `BEFORE_COMPLETION`, `AFTER_COMPLETION`, `AFTER_FAILURE` and `AFTER_SUCCESS` transactional observers are not implemented yet
 
+== Non-standard Features
+
+=== Qualified Injected Fields
+
+In CDI, if you declare a field injection point you need to use `@Inject` and optionally required qualifiers:
```
[the remark this answers is not in the document]
Comment (mkouba) [reply docs/src/main/asciidoc/cdi-reference.adoc]: Ok. `required qualifiers` is an official term but I agree that your proposal read much better...

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 3]
```diff
@@ -168,6 +170,155 @@ public class CounterBean {
...
+
+In Quarkus, you can skip the `@Inject` annotation completely if an injected field declares at least one qualifier:
+
+[source,java]
+----
+  @ConfigProperty(name = "cool")
+  String coolProperty;
+----
+
+NOTE: `@Inject` is still required for constructor and method injection.
```
[the remark this answers is not in the document]
Comment (mkouba) [reply docs/src/main/asciidoc/cdi-reference.adoc]: I don't think so. The point of this note is to clarify that `@Inject` is **still required** for constructors and initializers...

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 4]
```diff
@@ -168,6 +170,155 @@ public class CounterBean {
...
+by enabling additional logging via the following line in your `application.properties`.
+
+----
+quarkus.log.category."io.quarkus.arc.processor".level=DEBUG
+----
+
+[[default_beans]]
+=== Default Beans
+
+Quarkus adds a capability that CDI currently does not support which is to conditionally declare a bean if no such bean has been declared in the typical ways Quarkus supports.
```
[the remark this answers is not in the document]
Comment (mkouba) [reply docs/src/main/asciidoc/cdi-reference.adoc]: Well, I'm not the original author of these senteces... my explanation would be even simpler: https://quarkus.io/blog/quarkus-dependency-injection/#default-beans

**Comment by mkouba** [comment]
@manovotn would you care to propose changes directly in the PR?

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (manovotn) [reply docs/src/main/asciidoc/cdi-reference.adoc]: Later on we state - `Also if there's only one constructor there is no need for `@Inject`.`
So I just thought the combination of the two can be rather confusing.

**Comment by mkouba** [comment]
> [INFO] Quarkus - Integration Tests - Kafka ................ FAILURE [ 37.599 s]

and:

```
Caused by: java.util.NoSuchElementException
	at org.apache.kafka.common.utils.AbstractIterator.next(AbstractIterator.java:52)
	at io.quarkus.it.kafka.KafkaConsumerManager.receive(KafkaConsumerManager.java:40)
	at io.quarkus.it.kafka.KafkaConsumerManager_ClientProxy.receive(KafkaConsumerManager_ClientProxy.zig:51)
```
Anybody has a clue?

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkouba) [reply docs/src/main/asciidoc/cdi-reference.adoc]: Eh, yes I forgot about this special use case.

**Comment by manovotn** [comment]
> @manovotn would you care to propose changes directly in the PR?

Good point, I never seem to recall I can do that.

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 2 above)
Comment (manovotn) [inline docs/src/main/asciidoc/cdi-reference.adoc]:
```suggestion
In CDI, if you declare a field injection point you need to use `@Inject` and optionally a set of qualifiers:
```

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 3 above)
Comment (manovotn) [inline docs/src/main/asciidoc/cdi-reference.adoc]:
```suggestion
NOTE: With the notable exception of one special case discussed below, `@Inject` is still required for constructor and method injection.
```

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 4 above)
Comment (manovotn) [inline docs/src/main/asciidoc/cdi-reference.adoc]:
```suggestion
Quarkus adds a capability that CDI currently does not support which is to conditionally declare a bean if no other bean with equal types and qualifiers was declared by any available means (bean class, producer, synthetic bean, ...)
```

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 1 above)
Comment (manovotn) [reply docs/src/main/asciidoc/cdi-reference.adoc]: @mkouba apparently, I cannot suggest a multi line code change, but i would still delete this sentence as I think it just repeats what is said/implied from text and code above.

**Review by manovotn — COMMENTED** [review]
Tried adding the code suggestions.

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 5]
```diff
@@ -168,6 +170,155 @@ public class CounterBean {
...
+
+    @Produces
+    @DefaultBean
+    public Reporter reporter(){
+        // create a Reporter
+    }
+}
+----
+
+The idea is that the extension auto-configures things for user, eliminating a lot of boilerplate - we can just `@Inject` a `Tracer` wherever it is needed.
```
Comment (gsmet) [inline docs/src/main/asciidoc/cdi-reference.adoc]:
```suggestion
The idea is that the extension auto-configures things for the user, eliminating a lot of boilerplate - we can just `@Inject` a `Tracer` wherever it is needed.
```
?

**Review by gsmet — CHANGES_REQUESTED** [review]
I added one small comment. Agreed it should be backported.

**Review by manovotn — APPROVED** [review]

**Comment by mkouba** [comment]
@gsmet Could we merge this one?

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
@mkouba sorry, it took me some time but it's now in!

**Comment by gsmet** [comment]
Hum, I've seen it too late but next time, please squash everything, it's easier to backport. Thanks!

**Comment by mkouba** [comment]
Will do, sorry.

**Pull request merged**